src/verifiers/sev-snp/verify_evidence.c

```diff
@@ -0,0 +1,220 @@
+/* Copyright (c) 2022 Intel Corporation
+ * Copyright (c) 2020-2022 Alibaba Cloud
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <openssl/ssl.h>
+#include <openssl/ossl_typ.h>
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+#include <rats-tls/log.h>
+#include <rats-tls/verifier.h>
+#include "../../attesters/sev-snp/sev_snp.h"
+#include "sevapi.h"
+#include "x509cert.h"
+#include "crypto.h"
+
+int generate_vcek_pem(const uint8_t *chip_id, size_t chip_id_size, const snp_tcb_version_t *tcb)
+{
+	int cmd_ret = -1;
+	int count = 0;
+
+	/* Generate VCEK cert correspond to @chip_id and @tcb */
+	char cmdline_str[300] = {
+		0,
+	};
+	/* 2 chars per byte +1 for null term */
+	char id_buf[chip_id_size * 2 + 1];
+	memset(id_buf, 0, sizeof(id_buf));
+	for (uint8_t i = 0; i < chip_id_size; i++) {
+		sprintf(id_buf + 2 * i * sizeof(uint8_t), "%02x", chip_id[i]);
+	}
+	count = snprintf(cmdline_str, sizeof(cmdline_str), "wget -O %s \"%sMilan/%s",
+			 VCEK_DER_FILENAME, KDS_VCEK, id_buf);
+
+	char *TCBStringArray[8];
+	TCBStringArray[0] = "blSPL=";
+	TCBStringArray[1] = "teeSPL=";
+	TCBStringArray[2] = "reserved0SPL=";
+	TCBStringArray[3] = "reserved1SPL=";
+	TCBStringArray[4] = "reserved2SPL=";
+	TCBStringArray[5] = "reserved3SPL=";
+	TCBStringArray[6] = "snpSPL=";
+	TCBStringArray[7] = "ucodeSPL=";
+
+	count += snprintf((char *)&cmdline_str[count], sizeof(cmdline_str) - count,
+			  "?%s%02u&%s%02u&%s%02u&%s%02u\"", TCBStringArray[0],
+			  (unsigned)tcb->f.boot_loader, TCBStringArray[1], (unsigned)tcb->f.tee,
+			  TCBStringArray[6], (unsigned)tcb->f.snp, TCBStringArray[7],
+			  (unsigned)tcb->f.microcode);
+	cmdline_str[count] = '\0';
+
+	/* Don't re-download the VCEK from the KDS server if you already have it */
+	if (get_file_size(VCEK_DER_FILENAME) == 0) {
+		if (execute_system_command(cmdline_str)) {
+			RTLS_ERR("failed to download %s\n", VCEK_DER_FILENAME);
+			return cmd_ret;
+		}
+	}
+
+	cmd_ret = convert_der_to_pem(VCEK_DER_FILENAME, VCEK_PEM_FILENAME);
+	if (cmd_ret == -1)
+		RTLS_ERR("failed to convert %s to %s\n", VCEK_DER_FILENAME, VCEK_PEM_FILENAME);
+
+	return cmd_ret;
+}
+
+int generate_ark_ask_pem(void)
+{
+	int cmd_ret = -1;
+	int count = 0;
+	char cmdline_str[200] = {
+		0,
+	};
+	count = snprintf(cmdline_str, sizeof(cmdline_str), "wget --no-proxy -O %s %sMilan/%s",
+			 VCEK_CERT_CHAIN_PEM_FILENAME, KDS_VCEK, KDS_VCEK_CERT_CHAIN);
```

Comment:
if current workdir is readonly, how do you ensure you can write out the downloaded file successfully? Use mkdtemp() as a writable directory to hold the downloaded stuffs.

Reply:
Thx, I have add the check for {"openssl", "wget", "csplit" } in pre().

Due to the certs downloads from AMD KDS can be used for verifying the report multiple times. So I download the certs in `/opt/sev-snp` directory rather than temporary directory.